shadowsocks/dialer.go

```diff
@@ -0,0 +1,165 @@
+	proxyIP, err := net.ResolveIPAddr("ip", host)
+	if err != nil {
+		return nil, errors.New("Failed to resolve proxy address")
+	}
+	aead, err := newAeadCipher(cipher, password)
+	if err != nil {
+		return nil, err
+	}
+	d := ssDialer{proxyIP: net.ParseIP(proxyIP.String()), proxyPort: port, cipher: aead}
+	return &d, nil
+}
+
+func (d *ssDialer) DialTCP(address string) (onet.DuplexConn, error) {
+	proxyAddr := &net.TCPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	conn, err := net.DialTCP("tcp", nil, proxyAddr)
+	if err != nil {
+		return nil, err
+	}
+	ssw := NewShadowsocksWriter(conn, d.cipher)
+	socksTargetAddr := socks.ParseAddr(address)
+	if socksTargetAddr == nil {
+		return nil, errors.New("Invalid target address")
+	}
+	_, err = ssw.Write(socksTargetAddr)
+	if err != nil {
+		conn.Close()
+		return nil, errors.New("Failed to write target address")
+	}
+	ssr := NewShadowsocksReader(conn, d.cipher)
+	return onet.WrapConn(conn, ssr, ssw), nil
+}
+
+// Clients are encouraged to use io.ReadWriter methods of onet.PacketConn
+// to leverage the association with the proxy.
+func (d *ssDialer) DialUDP(address string) (onet.PacketConn, error) {
+	proxyAddr := &net.UDPAddr{IP: d.proxyIP, Port: d.proxyPort}
+	pc, err := net.ListenPacket("udp", "")
+	if err != nil {
+		return nil, err
+	}
+	targetHost, targetPortStr, err := net.SplitHostPort(address)
+	if err != nil {
+		return nil, errors.New("Invalid target address")
+	}
+	// Ignore the error, it would have failed when splitting the address
+	targetPort, _ := strconv.Atoi(targetPortStr)
+	targetAddr := &packetConnAddr{Host: targetHost, Port: targetPort}
+	conn := packetConn{
+		PacketConn: pc, proxyAddr: proxyAddr, targetAddr: targetAddr,
+			cipher: d.cipher, buf: make([]byte, udpBufSize)}
+	return &conn, nil
+}
+
+type packetConn struct {
+	net.PacketConn
+	io.ReadWriter
+	proxyAddr  *net.UDPAddr
+	targetAddr net.Addr
+	cipher     shadowaead.Cipher
+	m          sync.Mutex
+	buf        []byte // Write lock
+}
+
+func (c *packetConn) Write(b []byte) (int, error) {
+	return c.WriteTo(b, c.proxyAddr)
+}
+
+// WriteTo encrypts b and write to addr using the embedded PacketConn.
+func (c *packetConn) WriteTo(b []byte, addr net.Addr) (int, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+	socksTargetAddr := socks.ParseAddr(c.targetAddr.String())
+	if socksTargetAddr == nil {
+		return 0, errors.New("Invalid target address")
+	}
+	buf, err := shadowaead.Pack(c.buf, append(socksTargetAddr, b...), c.cipher)
+	if err != nil {
+		return 0, err
+	}
+	_, err = c.PacketConn.WriteTo(buf, addr)
+	return len(b), err
+}
+
+func (c *packetConn) Read(b []byte) (int, error) {
+	n, _, err := c.ReadFrom(b)
+	return n, err
+}
+
+// ReadFrom reads from the embedded PacketConn and decrypts into b.
+func (c *packetConn) ReadFrom(b []byte) (int, net.Addr, error) {
+	n, _, err := c.PacketConn.ReadFrom(b)
+	if err != nil {
+		return n, c.targetAddr, err
+	}
+	buf, err := shadowaead.Unpack(b[c.cipher.SaltSize():], b[:n], c.cipher)
```

Comment:
There's something off here. I don't see why you are skipping the salt size in the destination.
The input and destination should probably be a max size buffer, since b may not fit an encrypted packet.

Reply:
This is necessary so that the encrypted and destination buffers don't overlap (see [AEAD.Open](https://golang.org/pkg/crypto/cipher/#AEAD)). I have added a comment.